src/Mapbender/CoreBundle/Resources/public/mapbender.model.js

```diff
@@ -1084,118 +1043,67 @@ Mapbender.Model = {
-        changed.after = after;
-        changed.before = before;
+    setSourceLayerOrder: function(sourceId, newLayerIdOrder) {
+        var sourceIdx = {id: sourceId};
+        var sourceObj = this.getSource(sourceIdx);
+        var geoSource = Mapbender.source[sourceObj.type];
+
+        geoSource.setLayerOrder(sourceObj, newLayerIdOrder);
+
         this.mbMap.fireModelEvent({
             name: 'sourceMoved',
-            value: changed
+            // no receiver uses the bizarre "changeOptions" return value
+            // on this event
+            value: null
         });
         this._checkAndRedrawSource({
-            sourceIdx: {
-                id: source.id
-            },
-            options: {
-                children: {}
-            }
+            sourceIdx: sourceIdx,
+            options: {children: {}}
         });
     },
     /**
+     * Bring the sources identified by the given ids into the given order.
+     * All other sources will be left alone!
      *
+     * @param {string[]} newIdOrder
      */
-    _moveSource: function(source, before, after) {
-        var old_pos = this.getSourcePos(source);
-        var new_pos;
-        if (before && before.source && after && after.source) {
-            var before_pos = this.getSourcePos(before.source);
-            var after_pos = this.getSourcePos(after.source);
-            if (old_pos <= before_pos)
-                new_pos = before_pos;
-            else if (old_pos > before_pos)
-                new_pos = after_pos;
-        } else if (before && before.source) {
-            new_pos = this.getSourcePos(before.source);
-        } else if (after && after.source) {
-            new_pos = this.getSourcePos(after.source);
-        }
-        if (old_pos === new_pos)
-            return;
-        this.sourceTree.splice(new_pos, 0, this.sourceTree.splice(old_pos, 1)[0]);
-        var mqL = this.map.layersList[source.mqlid];
-        if (old_pos > new_pos) {
-            mqL.down(Math.abs(old_pos - new_pos));
-        } else {
-            mqL.up(Math.abs(old_pos - new_pos));
-        }
-        var changed = this.createChangedObj(source);
-        changed.after = after;
-        changed.before = before;
-        this.mbMap.fireModelEvent({
-            name: 'sourceMoved',
-            value: changed
+    reorderSources: function(newIdOrder) {
+        var self = this;
+        var sourceObjs = $.map(newIdOrder, function(sourceId) {
+            return self.findSource({id: sourceId});
+        });
+        // Collect currently set positions and z indexes for given sources.
+        // position := array index in this.sourceTree
+        // z index := mapquery layer position = openlayers map layer index - 1
+        // The collected values will be reused / redistributed to the affected
+        // sources.
+        var oldPositions = [];
+        var zIndexes = [];
+        var sourceIdToSource = {};
+        $.each(sourceObjs, function(i, sourceObj) {
```

Comment:
For better performance, it recommends using underscore/lodash or better native functions if no DOM is involved

Reply:
[Done](https://github.com/mapbender/mapbender/pull/852/commits/bd838ead0b4c8de5dca79a8136b32fe53148f31e)